File: Knights_Travails.js
function knightMoves(start, end) {
    // Define the possible moves a knight can make
    const moves = [
        [-2, -1], [-2, 1], [2, -1], [2, 1],
        [-1, -2], [-1, 2], [1, -2], [1, 2]
    ];

    // Helper function to check if a position is valid on the board
    function isValid(x, y) {
        return x >= 0 && y >= 0 && x < 8 && y < 8;
    }

    // Perform BFS to find the shortest path
    function bfs(start, end) {
        const queue = [[start]];
        const visited = new Set([start.toString()]);

        while (queue.length > 0) {
            const path = queue.shift();
            const [x, y] = path[path.length - 1];

            if (x === end[0] && y === end[1]) {
                return path;
            }

            for (const move of moves) {
                const newX = x + move[0];
                const newY = y + move[1];
                if (isValid(newX, newY)) {
                    const newPath = path.concat([[newX, newY]]);
                    const key = newPath[newPath.length - 1].toString();
                    if (!visited.has(key)) {
                        queue.push(newPath);
                        visited.add(key);
                    }
                }
            }
        }

        // If no path is found
        return null;
    }

    // Call BFS to find the shortest path
    const shortestPath = bfs(start, end);

    // Output the result
    if (shortestPath) {
        console.log(`You made it in ${shortestPath.length - 1} moves! Here's your path:`);
        for (const pos of shortestPath) {
            console.log(pos);
        }
        return shortestPath;
    } else {
        console.log("No path found.");
        return null;
    }
}

// Test cases
knightMoves([0,0],[1,2]);
knightMoves([0,0],[3,3]);
knightMoves([3,3],[0,0]);
knightMoves([0,0],[7,7]);
knightMoves([3,3],[4,3]);

File: knights_travails.rb
# Define class representing a chessboard
class Chessboard
  def initialize
    @board = Array.new(8) { Array.new(8) }
    populate_board
  end

  # Method to populate the board with coordinates
  def populate_board
    (0..7).each do |row|
      (0..7).each do |col|
        @board[row][col] = [row, col]
      end
    end
  end

  # Method to check if a position is valid on the board
  def valid_position?(position)
    row, col = position
    row >= 0 && row < 8 && col >= 0 && col < 8
  end

  # Method to find possible moves for a knight
  def possible_moves(position)
    moves = [[2, 1], [2, -1], [-2, 1], [-2, -1], [1, 2], [1, -2], [-1, 2], [-1, -2]]
    possible_moves = []

    moves.each do |move|
      new_row = position[0] + move[0]
      new_col = position[1] + move[1]
      possible_moves << [new_row, new_col] if valid_position?([new_row, new_col])
    end

    possible_moves
  end

  # Method to find the shortest path using breadth-first search
  def knight_moves(start, target)
    queue = [[start]]
    visited = []

    until queue.empty?
      current_path = queue.shift
      current_position = current_path.last

      return current_path if current_position == target

      possible_moves(current_position).each do |move|
        next if visited.include?(move)

        new_path = current_path + [move]
        queue << new_path
        visited << move
      end
    end
  end
end

# Driver script
def display_path(path)
  puts "You made it in #{path.length - 1} moves! Here's your path:"
  path.each { |position| p position }
end

board = Chessboard.new
start = [3, 3]
target = [4, 3]
path = board.knight_moves(start, target)
display_path(path)
